solutions/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,200 @@
+---
+id: izho-17-LongestBeautifulSequence
+source: IZhO 2017
+title: 2017 - Longest beautiful sequence
+author: Shijie Ren
+---
+
+<Spoiler title="Hint 1">
+
+Both $\mathcal{O}(N^2)$ and $\mathcal{O}(N\max{a[i]})$ algorithm is not hard to
+come up with,
+
+but the latter one is more enlightening.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+You may find two ways of $\mathcal{O}(N\max{a[i]})$ DP.
+
+One is to traverse all the situations of the previous number (kind of
+preprocessing previous answers), $\mathcal{O}(\max{a[i]})$ to query current
+answer in DP array and $\mathcal{O}(1)$ to update.
+
+The other is to traverse all the possible situations of the next number , (kind
+of predicting next answers) $\mathcal{O}(1)$ to query current answer in DP array
+and $\mathcal{O}(\max{a[i]})$ to update , but need a bit more memory.
+
+These algorithms are not efficient enough because the time complexity of update
+and query are unbalanced.
+
+Try to Consider the two equations carefully and try to "combine" their ideas.
+
+</Spoiler>
+
+<Spoiler title="Explanation">
+
+$\mathcal{O}(N^2)$ DP:
+
+Let $\texttt{dp}[i]$ be the length of longest beautiful sequence end with
+$a[i]$.
+
+$$
+\texttt{dp}[i]=\max_{j<i,\text{popcount}(a[i]\&a[j])=k[i]}\{\texttt{dp}[j]\}+1
+$$
+
+We can also use $a[i]$ as indices:
+
+$\mathcal{O}(N\max a[i])$ DP:
+
+Let $\texttt{dp}[i][j]$ be the length of longest beautiful sequence of first $i$
+elements end with $j$ (i.e. $j=\texttt{EndElement}$) ,
+
+then
+$\max\limits_{j\leq\max{a[i]},\text{popcount}(j\&a[i])=k[i]}\{\texttt{dp}[i-1][j]\}+1$
+will be the length of longest beautiful sequence end with $a[i]$.
+
+$$
+\texttt{dp}[i][a[i]] = \max \{ \texttt{dp}[i-1][a[i]] , \max_{j\leq\max{a[i]},\text{popcount}(j\&a[i])=k[i]}\{\texttt{dp}[i-1][j]\}+1 \}
+$$
+
+$$
+\texttt{dp}[i][j] = \texttt{dp}[i-1][j] , j\not=a[i]
+$$
+
+Or another way:
+
+Let $\texttt{dp}[i][j][q]$ be the length of longest beautiful sequence of first
+$i$ elements satisfied $\text{popcount}(\texttt{EndElement}\&j)=q$ ,
+
+then $\texttt{dp}[i-1][a[i]][k[i]]+1$ will be the length of longest beautiful
+sequence end with $a[i]$.
+
+$$
+\texttt{dp}[i][j][\text{popcount}(a[i]\&j)]=\max_{j\leq\max{a[i]}}\{\texttt{dp}[i-1][j][\text{popcount}(a[i]\&j)],\texttt{dp}[i-1][a[i]][k[i]]+1\}
+$$
+
+$$
+\texttt{dp}[i][j][q]=\texttt{dp}[i-1][j][q],q\not=\text{popcount}(a[i]\&j)
+$$
+
+According to Hint 2 , we need to consider either of them , that is , to consider
+the answers before and after.
+
+Since $a[i] < 2^{20}$ , divide the 20-bits number into two 10-bits numbers.
+
+For convenience , let $\text{b0}(x) = \lfloor \frac{x}{2^{10}} \rfloor$
+represents 10-19 bits , let $\text{b1}(x) = x \& 2^{10} $ represents 0-9 bits.

```

Comment:
```suggestion
According to Hint 2, we need to consider either of them, that is, to consider
the answers before and after.

Since $a[i] < 2^{20}$ , divide the 20-bits number into two 10-bits numbers.

For convenience, let $\text{b0}(x) = \lfloor \frac{x}{2^{10}} \rfloor$
represents 10-19 bits, let $\text{b1}(x) = x \& 2^{10} $ represents 0-9 bits.

```